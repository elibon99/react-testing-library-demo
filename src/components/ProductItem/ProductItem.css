
.product-item-container {
    display: flex;
    flex-direction: column;
    width: 200px;
    align-self: flex-end;

}

.product-item-image {
    display: block;
    max-width: 200px;
    max-height: 180px;
    width: auto;
    height: auto;
}

.product-item-title {
    font-size: 16px;
    font-weight: 600;
    color: #424242;
    margin: 0 0;
    margin-top: 12px;
}

.product-item-price {
    font-size: 14px;
    color: #424242;
    font-weight: normal;
    margin: 0 0;
    margin-top: 12px;
    margin-bottom: 20px;
}

.product-item-allowed {
    cursor: pointer;
    border: none;
    padding: 8px 8px;
    color: white;
    border-radius: 1000px;
    background: #18a0fb;
}

.product-item-btn {
    cursor: pointer;
    border: none;
    padding: 8px 8px;
    color: white;
    border-radius: 1000px;
    background: #18a0fb;
}

.product-item-btn-disabled {
    background-color: rgba(24,160,251, 0.3);
    cursor: not-allowed;
}

.product-item-disabled {
    cursor: pointer;
    border-width: 1px;
    background-color: red;
    color: white;
}