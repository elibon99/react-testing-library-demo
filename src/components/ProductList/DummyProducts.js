import Macbook from "../../assets/images/Macbook.jpg"
import Iphone from "../../assets/images/Iphone.jpg"
import AirPods from "../../assets/images/AirPods.webp"

export const DummyProducts = [
    {
        id: 1,
        title: "Macbook",
        price: 16000,
        image: Macbook
    },
    {
        id: 2,
        title: "Iphone",
        price: 7000,
        image: Iphone
    },
    {
        id: 3,
        title: "AirPods",
        price: 1500,
        image: AirPods
    }
]