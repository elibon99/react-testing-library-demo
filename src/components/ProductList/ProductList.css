
.product-list-container {
    display: flex;
    flex-direction: column;
    background: white;
    border-radius: 8px;
    width: 65%;
    padding: 30px 30px;
}

.product-list {
    display: flex;
    justify-content: space-between;
    margin-top: 20px;
    align-items: stretch;
}

.product-list-header {
    margin: 0 0;
}