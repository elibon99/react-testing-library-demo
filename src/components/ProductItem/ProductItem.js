import React from 'react'
import "./ProductItem.css"

const ProductItem = ({product, products, addProductToCart}) => {
    const isProductInShoppingCart = products.find((pr) => pr.id === product.id)
    return (
        <div className="product-item-container" id={"product-item"}>
            <img className="product-item-image" src={product.image} alt={"product"}/>
            <h4 className="product-item-title">{product.title}</h4>
            <h4 className="product-item-price">{`${product.price} SEK`}</h4>
            <button
                title={product.title}
                onClick={() => !isProductInShoppingCart && addProductToCart(product.id)}
                className={`product-item-btn ${isProductInShoppingCart && "product-item-btn-disabled"}`}
            >
                Add to cart
            </button>
        </div>
    )
}

export default ProductItem