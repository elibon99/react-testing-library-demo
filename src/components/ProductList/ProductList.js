import React from 'react'
import "./ProductList.css"
import {DummyProducts} from "./DummyProducts";
import ProductItem from "../ProductItem/ProductItem";

const ProductList = ({products, setProducts}) => {
    const addProductToCart = (id) => {
        let product = DummyProducts.find((product) => product.id === id)
        setProducts(prev => [...prev, product])
    }

    return (
        <div className="product-list-container">
            <h3 className="product-list-header">Products</h3>
            <div className="product-list" data-testid={"product-list"}>
                {DummyProducts.map((product, index) => {
                    return(
                        <ProductItem key={index} product={product} products={products} addProductToCart={addProductToCart}/>
                    )
                })}
            </div>
        </div>
    )
}

export default ProductList