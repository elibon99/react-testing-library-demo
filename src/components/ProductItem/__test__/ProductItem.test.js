import {render, screen} from "@testing-library/react";
import ProductItem from "../ProductItem";
import {DummyProducts} from "../../ProductList/DummyProducts";

const ShoppingCart = DummyProducts.filter((product) => product.title === "Macbook")
const Macbook = ShoppingCart[0]

describe('<ProductItem/>', () => {

    /**
     * Checks that ProductItem is rendered correctly by comparing with snapshot
     */
    it('renders correctly', async () => {
        const {asFragment} = render(
            <ProductItem
                product={Macbook}
                products={ShoppingCart}
                addProductToCart={() => {}}
            />
            )
        expect(asFragment()).toMatchSnapshot()
    })

    /**
     * When product is in cart, add button should be disabled
     */
    it('add button should have class disabled when product is in cart', async () => {
        render(
            <ProductItem
                product={Macbook}
                products={ShoppingCart}
                addProductToCart={() => {}}
            />
        )

        const buttonElement = screen.getByRole("button")
        expect(buttonElement).toHaveClass("product-item-btn-disabled")
    })
})
