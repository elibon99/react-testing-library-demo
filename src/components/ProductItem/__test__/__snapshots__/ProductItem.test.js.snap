// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductItem/> renders correctly 1`] = `
<DocumentFragment>
  <div
    class="product-item-container"
    id="product-item"
  >
    <img
      alt="product"
      class="product-item-image"
      src="Macbook.jpg"
    />
    <h4
      class="product-item-title"
    >
      Macbook
    </h4>
    <h4
      class="product-item-price"
    >
      16000 SEK
    </h4>
    <button
      class="product-item-btn product-item-btn-disabled"
      title="Macbook"
    >
      Add to cart
    </button>
  </div>
</DocumentFragment>
`;
